compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/StringTrimLowering.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.common.lower
+
+import org.jetbrains.kotlin.backend.common.CommonBackendContext
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.lower.matchers.SimpleCalleeMatcher
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrConst
+import org.jetbrains.kotlin.ir.expressions.IrConstKind
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.isString
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.FqName
+
+val computeStringTrimPhase = makeIrFilePhase(
+    ::StringTrimLowering,
+    name = "StringTrimLowering",
+    description = "Compute trimIndent and trimMargin operations on constant strings"
+)
+
+class StringTrimLowering(val context: CommonBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid(this)
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        return when {
+            trimIndentMatcher(expression) -> maybeComputeTrimIndent(expression)
+            trimMarginMatcher(expression) -> maybeComputeTrimMargin(expression)
+            else -> super.visitCall(expression)
+        }
+    }
+
+    private fun maybeComputeTrimIndent(call: IrCall): IrExpression {
+        val receiverString = call.extensionReceiver!!.getConstantString() ?: return call
```

Comment:
Seems it's also possible to make partial removal of margins and indents if we have constant entries at beginning and ending in IrStringConcatenation

Reply:
I'm not sure this is safe. It also leads to complex rewrites where you have to move the trimIndent/Margin call onto the non-constant replacement characters.

Consider
```kotlin
fun foo(a: String): String {
  return """
    indented by 4
      indented by 6
$a
    indented by 4
    """.trimIndent()
}
```
If `a` has an indent of 2 spaces then only 2 spaces should be removed from the prefix/suffix constants. If `a` has an indent of 8 spaces then 4 spaces will correctly be removed from the prefix/suffix constants, but `a` itself will have an indent of 8 spaces which will be removed instead of only 4.